File: static/SocketRequests.js
class Message
{
    constructor(messageKey)
    {
        this.messageKey = messageKey
        this.message = {}
        this._messageContent = {}
    }

    _setMessage(data)
    {
        this.message = data
    }

    setMessage(data)
    {
        this._setMessage(data)
        this._messageContent = this.message
    }

    get messageContent() {
        return this._messageContent;
    }
}


// ------------------ downloadMedia -------------------
class DownloadMedia
{
    constructor(link, type)
    {
        this._link = link;
        this._type = type;
    }

    get link()
    {
        return this._link;
    }

    get type()
    {
        return this._type;
    }
}

class DownloadMediaRequest extends Message
{
    constructor()
    {
        super("downloadMedia");
    }

    _setMessage(data)
    {
        if (!(data instanceof DownloadMedia))
        {
            console.error("wrong data for message")
            return
        }
        this.message["link"] = data.link
        this.message["type"] = data.type
    }
}


// ------------------ downloadPlaylists -------------------
class DownloadPlaylistsRequest extends Message
{
    constructor()
    {
        super("downloadPlaylists");
    }
}


// ------------------ archiveSong -------------------
class ArchiveSong
{
    constructor(ytHash, index, playlistName)
    {
        this.ytHash = ytHash;
        this.index = index;
        this.playlistName = playlistName;
    }
}

class ArchiveSongRequest extends Message
{
    constructor()
    {
        super("archiveSong")
    }

    _setMessage(data)
    {
        if (!(data instanceof ArchiveSong))
        {
            console.error("wrong data for message")
            return
        }
        this.message["ytHash"] = data.ytHash
        this.message["index"] = data.index
        this.message["playlistName"] = data.playlistName
    }
}

File: Common/SocketRequests.py
# ------------------------- downloadMedia -------------------------
class DownloadMedia:
    def __init__(self, link, type):
        self.link = link
        self.type = type

class DownloadMediaRequest:
    message = "downloadMedia"

    def __init__(self, msg):
        self.downloadMedia = DownloadMedia(msg["link"], msg["type"])


# --------------------- downloadPlaylists ----------------------------
class DownloadPlaylistsRequest:
    message = "downloadPlaylists"


# ------------------------ archiveSong -------------------------------
class ArchiveSong:
    def __init__(self, ytHash, index, playlistName):
        self.ytHash = ytHash
        self.index = index
        self.platlistName = playlistName

class ArchiveSongRequest:
    message = "archiveSong"
    def __init__(self, msg):
        self.archiveSong = ArchiveSong(msg["ytHash"], msg["index"], msg["playlistName"])
